File: python/src/hdz_auth/jwt.py
from __future__ import annotations

import base64
import json


def extract_user_id_from_jwt(token: str) -> str | None:
    """Extract Supabase user id (sub) from JWT without verification."""
    try:
        parts = token.split(".")
        if len(parts) != 3:
            return None
        padded = parts[1] + "=" * (-len(parts[1]) % 4)
        payload = json.loads(base64.urlsafe_b64decode(padded.encode()))
        return payload.get("sub")
    except Exception:
        return None

File: ts/src/jwt.ts
/**
 * Extract Supabase user id (sub claim) from a JWT access token.
 * Works in browser, Workers, and Node without external deps.
 */
export function extractUserIdFromJwt(token: string): string | null {
  try {
    const parts = token.split('.')
    if (parts.length !== 3) return null
    const payloadJson = atob(parts[1].replace(/-/g, '+').replace(/_/g, '/'))
    const payload = JSON.parse(payloadJson) as { sub?: string }
    return payload.sub ?? null
  } catch {
    return null
  }
}

export function isJwtExpired(token: string, skewSeconds = 30): boolean {
  try {
    const parts = token.split('.')
    if (parts.length !== 3) return true
    const payload = JSON.parse(atob(parts[1].replace(/-/g, '+').replace(/_/g, '/'))) as { exp?: number }
    if (!payload.exp) return false
    return payload.exp <= Math.floor(Date.now() / 1000) + skewSeconds
  } catch {
    return true
  }
}
